### PR TITLE
refactor: using aether consumer sdk

### PR DESCRIPTION
**Pull request by lordmallam**
Implements CKAN Consumer using updated Consumer SDK

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
-script: sh bin/travis-run.sh
 
-deploy:
-- provider: script
-  script: "./scripts/release.sh"
-  skip_cleanup: true
-  on:
-    all_branches: true
-    condition: ($TRAVIS_BRANCH =~ ^release\-[0-9]+\.[0-9]+[\.0-9]*$) || ($TRAVIS_TAG =~ ^[0-9]+\.[0-9]+[\.0-9]*$)
+matrix:
```
Comment (obdulia-losantos) [inline .travis.yml]: `matrix` was renamed to `jobs` :grimacing:

File: .travis.yml [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
+      stage: test
+      install: true
+      script: "./scripts/run_travis.sh"
+
+    - name: "Release"
+      stage: deploy
+      install: true
+      script: "./scripts/release.sh"
+      # release only if tag like #.#.#
+      if: (tag =~ ^[0-9]+(\.[0-9]+){2}$)
```
Comment (obdulia-losantos) [inline .travis.yml]: We need to build only PR and the master branch or tags.

```yml
# build only the master branch or tags like #.#.#
branches:
  only:
    - master
    - /^[0-9]+(\.[0-9]+){2}$/

# define stages and their execution order
stages:
  - name: test

  # release only in:
  #   - branch   master
  #   - tag      #.#.#
  #   - never in forks or pull requests
  - name: release
    if: |
      fork IS false AND \
      type != pull_request AND \
      ((branch = master) OR \
      (tag =~ ^[0-9]+(\.[0-9]+){2}$))

  # Trigger deployment on GCS
  # only in:
  #   - branch   master
  #   - tag      #.#.#
  #   - never in forks or pull requests
  - name: deploy
    if: |
      fork IS false AND \
      type != pull_request AND \
      ((branch = master) OR \
      (tag =~ ^[0-9]+(\.[0-9]+){2}$))

# use jobs to parallelize tests
jobs:
  fast_finish: true
  include:
    - name: "Test"
      stage: test
      script: "./scripts/run_travis.sh"

    - name: "Release"
      stage: release
      script: ./scripts/release.sh

    - name: "Deployment"
      stage: deploy
      script: ./scripts/deploy.sh

install: true
```

File: consumer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,36 @@
+FROM python:3.7-slim-stretch
```
Comment (obdulia-losantos) [inline consumer/Dockerfile]: now it's `buster` :wink: 

```
FROM python:3.7-slim-buster
```

File: consumer/app/__init___.py [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
-#
+#!/usr/bin/env python
+
 # Copyright (C) 2018 by eHealth Africa : http://www.eHealthAfrica.org
 #
 # See the NOTICE file distributed with this work for additional information
 # regarding copyright ownership.
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed under the Apache License, Version 2.0 (the 'License');
```
Comment (obdulia-losantos) [inline consumer/app/__init___.py]:
```suggestion
# Licensed under the Apache License, Version 2.0 (the "License");
```

File: consumer/app/__init___.py [hunk 5]
```diff
@@ -1,11 +1,11 @@
-#!/usr/bin/env bash
-#
+#!/usr/bin/env python
```
Comment (obdulia-losantos) [inline consumer/app/__init___.py]: it this line needed?

File: consumer/app/artifacts.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
+#!/usr/bin/env python
```
Comment (obdulia-losantos) [inline consumer/app/artifacts.py]: this line is required only if you execute the file.

File: consumer/app/__init___.py [hunk 7]
```diff
@@ -1,11 +1,11 @@
...
+
 # Copyright (C) 2018 by eHealth Africa : http://www.eHealthAfrica.org
```
Comment (obdulia-losantos) [inline consumer/app/__init___.py]: Maybe it's time to upgrade to `2020` :thinking:

File: .travis.yml [hunk 8]
```diff
@@ -1,18 +1,24 @@
 language: python
 python:
-  - "2.7"
```
Comment (shawnsarwar) [inline .travis.yml]: I remember there was a reason that the Viderum guys built this on 2.x.
There may be some incompatibility in the CKAN python libraries. Just something to watch out for.

File: consumer/app/config.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            kafka_security[i.lower()] = kafka_security[i]
+            del kafka_security[i]
+    return kafka_security
+
+
+def get_consumer_config():
+    return consumer_config
+
+
+load_config()
```
Comment (obdulia-losantos) [inline consumer/app/config.py]: :thinking:

File: consumer/app/artifacts.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+                self.definition.user,
+                self.definition.password
+            )
+        except AttributeError:
+            pass  # may not need creds
+        # add an _id so we can check the instance
+        setattr(self.session, 'instance_id', str(uuid4()))
+        return self.session
+
+    @lock
```
Comment (shawnsarwar) [inline consumer/app/artifacts.py]: We only need to @lock this method if having multiple requests at the same time isn't safe to do in parallel. I've probably overused it in the past, but something to remember.

File: consumer/conf/pip/primary-requirements.txt [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+eha_jsonpath
+flake8
+flake8-quotes
+jsonschema
+pytest
+pytest-cov
+pytest-runner
+requests
+responses
+sqlalchemy
```
Comment (obdulia-losantos) [inline consumer/conf/pip/primary-requirements.txt]: Could you split the requirements in the "consumer" ones and the "tests" ones?

File: consumer/entrypoint.sh [hunk 12]
```diff
@@ -0,0 +1,101 @@
+#!/bin/bash
```
Comment (obdulia-losantos) [inline consumer/entrypoint.sh]: License header

File: consumer/setup.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+# under the License.
+
+from setuptools import setup
+setup(
+    name='aether_ckan_consumer',
+    author='eHealth Africa',
+    author_email='aether@ehealthafrica.org',
+    decription='Aether CKAN Kafka consumer',
+    version='2.0.0',
+    setup_requires=['pytest-runner'],
```
Comment (obdulia-losantos) [inline consumer/setup.py]: really?

File: consumer/setup.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    tests_require=['pytest', 'pytest-cov'],
```
Comment (obdulia-losantos) [inline consumer/setup.py]: same comment

File: .travis.yml (hunk 8 above)
Comment (obdulia-losantos) [reply .travis.yml]: https://www.bleepingcomputer.com/news/software/python-27-reaches-end-of-life-after-20-years-of-development/

File: consumer/app/fixtures/examples.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    'name': 'Test Subscription',
+    'topic_pattern': '*',
+    'topic_options': {
+        'masking_annotation': '@aether_masking',    # schema key for mask level of a field
+        'masking_levels': ['public', 'private'],    # classifications
+        'masking_emit_level': 'public',             # emit from this level ->
+        'filter_required': True,                    # filter on a message value?
+        'filter_field_path': 'operational_status',  # which field?
+        'filter_pass_values': ['operational'],      # what are the passing values?
+    }
```
Comment (shawnsarwar) [inline consumer/app/fixtures/examples.py]: We might want to let people add some options about the target within CKAN as a section here.  For example, this subscription is going to write to a dataset. Do we want to let them name the dataset? 
```json
"target_options": {
    "name": "a name for the dataset",
    "default_read_level": "public"
}
```
Are there other options for datasets that we want to specify here? You can keep it very simple for now.

File: consumer/conf/consumer/consumer.json [hunk 16]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (shawnsarwar) [inline consumer/conf/consumer/consumer.json]: This should no longer be picked up from the file system. All jobs / resource should be created via API. You can make this a couple of test artifacts though and use the API to start them up there.

File: consumer/conf/consumer/consumer.json [hunk 17]
```diff
@@ -0,0 +1,7 @@
...
+    "title": "CKAN portal",
+    "url": "http://ckan:5000",
+    "api_key": "49d4cd30-85ab-467f-8710-676f0ce55df1",
+    "autoconfig_datasets" : true,
```
Comment (shawnsarwar) [inline consumer/conf/consumer/consumer.json]: We can kill autoconfig totally.

A adding a subscription to `topics : "*"` is the same as what the old autoconfig montior would handle.

Comment (lordmallam) [reply consumer/conf/consumer/consumer.json]: `autoconfig`  in the current implementation creates a 1:1 topics:datasets, won't this be the default option if a target dataset isn't specified in `target_options`? Making a mandatory `target_options` will mean the user has to have setup datasets on ckan before hand.

File: consumer/conf/pip/primary-requirements.txt [hunk 18]
```diff
@@ -8,21 +8,24 @@
 # To check the outdated dependencies run                                                        #
 #       docker-compose run --rm --no-deps ckan-consumer eval pip list --outdated                #
 # Update me and then run                                                                        #
-#       docker-compose run --rm ckan-consumer pip_freeze                                             #
+#       docker-compose run --rm ckan-consumer pip_freeze                                        #
 #                                                                              					#
 #################################################################################################
```
Comment (obdulia-losantos) [inline consumer/conf/pip/primary-requirements.txt]: :thinking: 

image

File: consumer/setup.py (hunk 14 above)
Comment (obdulia-losantos) [reply consumer/setup.py]: `tests_require` ...

flake8
flake8-quotes
jsonschema
pytest
pytest-cov
pytest-runner

File: consumer/setup.py [hunk 19]
```diff
@@ -25,7 +25,16 @@
     author_email='aether@ehealthafrica.org',
     decription='Aether CKAN Kafka consumer',
     version='2.0.0',
```
Comment (obdulia-losantos) [inline consumer/setup.py]: fixed value or maybe we can use the tags???

File: consumer/app/__init___.py (hunk 4 above)
Comment (obdulia-losantos) [reply consumer/app/__init___.py]: This change applies to **ALL** the license headers :wink:

File: .travis.yml [hunk 20]
```diff
@@ -1,18 +1,44 @@
...
+jobs:
+  fast_finish: true
+  include:
+
+    - name: "Test"
+      stage: test
+      script: "./scripts/run_travis.sh"
+
+    - name: "Release"
+      stage: deploy
```
Comment (obdulia-losantos) [inline .travis.yml]:
```suggestion
      stage: release
```

It's trying to deploy :scream: 
https://travis-ci.com/eHealthAfrica/aether-ckan-consumer/builds/146917234

File: .gitignore [hunk 21]
```diff
@@ -1,3 +1,5 @@
 *.pyc
 /db
 .vscode
+/ckan
+*.coverage
```
Comment (obdulia-losantos) [inline .gitignore]: The `consumer/.coverage` file was committed to the repo :woman_facepalming:

File: consumer/tests/ckan/Dockerfile [hunk 22]
```diff
@@ -1,5 +1,9 @@
+# ----------------------------------------------------
+# https://github.com/ckan/ckan/blob/master/Dockerfile
+# ----------------------------------------------------
+
 # See CKAN docs on installation from Docker Compose on usage
-FROM debian:jessie
+FROM debian:stretch
```
Comment (obdulia-losantos) [inline consumer/tests/ckan/Dockerfile]: not `buster`?

File: docker-compose-test.yml [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+      REDIS_PASSWORD: testpassword
+
+      KAFKA_URL: kafka-test:29092
+      SECURITY.PROTOCOL: SASL_SSL
+      SASL.MECHANISM: PLAIN
+      SASL.USERNAME: root
+      SASL.PASSWORD: 09049a938686107c15ed6748010e2f1b
+
+  ckan-test:
+    image: ehealthafrica/ckan:2.8.0-rc
```
Comment (obdulia-losantos) [inline docker-compose-test.yml]: are you using the uploaded image or the one you are building? :thinking:

File: consumer/tests/ckan/.env [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+CKAN_PORT=5000
+#
+# Email settings
+CKAN_SMTP_SERVER=smtp.corporateict.domain:25
+CKAN_SMTP_STARTTLS=True
+CKAN_SMTP_USER=user
+CKAN_SMTP_PASSWORD=pass
+CKAN_SMTP_MAIL_FROM=ckan@ckan-test
+
+DATASTORE_READONLY_PASSWORD=datastore
```
Comment (obdulia-losantos) [inline consumer/tests/ckan/.env]: :thinking:

File: docker-compose-test.yml (hunk 23 above)
Comment (lordmallam) [reply docker-compose-test.yml]: uploaded

File: consumer/tests/ckan/.env (hunk 24 above)
Comment (lordmallam) [reply consumer/tests/ckan/.env]: The CKAN setup is only used for integration testing

File: docker-compose-test.yml (hunk 23 above)
Comment (obdulia-losantos) [reply docker-compose-test.yml]: sorry, it was confusing, you were removing the files, not adding them :face_with_head_bandage:

File: .travis.yml (hunk 8 above)
Comment (obdulia-losantos) [reply .travis.yml]: In the README says python 2.7  :nerd_face: 

> ### Running the app
> The project is setup with Docker. It will spin up an environment with Python **2.7** based on Alpine and install the required dependencies.

File: consumer/tests/ckan/Dockerfile [hunk 25]
```diff
@@ -3,7 +3,7 @@
 # ----------------------------------------------------
 
 # See CKAN docs on installation from Docker Compose on usage
-FROM debian:stretch
+FROM debian:buster
```
Comment (obdulia-losantos) [inline consumer/tests/ckan/Dockerfile]: One stupid thing, if you are using the uploaded image... why is this here?

Comment (lordmallam) [reply consumer/tests/ckan/Dockerfile]: From the viderum setup, dockerfile configures the image. Its usage is unclear right now. It will be streamlined to our needs after the consumer job is defined and implemented. https://jira.ehealthafrica.org/browse/AET-595

Comment (obdulia-losantos) [reply consumer/tests/ckan/Dockerfile]: _lost_ :face_with_head_bandage:

Comment (lordmallam) [reply consumer/tests/ckan/Dockerfile]: or was this done by us? https://github.com/eHealthAfrica/aether-bootstrap/blob/master/ckan/Dockerfile

Comment (obdulia-losantos) [reply consumer/tests/ckan/Dockerfile]: I think this is a question for @shawnsarwar

File: .travis.yml (hunk 8 above)
Comment (shawnsarwar) [reply .travis.yml]: @lordmallam I assume this is all ok.

File: consumer/app/fixtures/examples.py (hunk 15 above)
Comment (lordmallam) [reply consumer/app/fixtures/examples.py]: include the `target_options` that allows users define target dataset. Dataset is created if does not exist.
```json
"target_options": {
        "dataset_metadata": {
            "title": "Pollution in Nigeria",
            "name": "pollution-in-nigeria111",
            "owner_org": "eHA",
            "notes": "Some description",
            "author": "eHealth Africa",
            "private": False
        }
    }
```

File: consumer/conf/consumer/consumer.json (hunk 17 above)
Comment (lordmallam) [reply consumer/conf/consumer/consumer.json]: Now, topics will be added to the specified target dataset as resources.

File: consumer/tests/ckan/Dockerfile (hunk 25 above)
Comment (lordmallam) [reply consumer/tests/ckan/Dockerfile]: @obdulia-losantos it is necessary to setup the datastore and datapusher https://docs.ckan.org/en/2.8/maintaining/installing/install-from-docker-compose.html#datastore-and-datapusher

Comment (obdulia-losantos) [reply consumer/tests/ckan/Dockerfile]: :woman_shrugging:
